python/tvm/te/schedule.py

```diff
@@ -519,9 +523,149 @@ def rolling_buffer(self):
+            # is a transpose of ``A``'s layout.  The loop iteration
+            # that computes ``B`` will correspond to ``B``'s memory
+            # layout.
+
+            A = te.placeholder([n,m])
+            B = te.compute(A.shape, lambda i,j: A[i,j])
+            s = te.create_schedule(B.op)
+
+            s[B].transform_layout(lambda i,j: [j,i])
+
+        """
+
+        args = []
+        var_arg_name = None
+        kwargs = collections.OrderedDict()
+        default_index_dtype = "int32"
+
+        # Make a dummy variable for each explicitly named input index.
+        # We may have some keyword-only arguments, if the function has
+        # *args before the last argument.
+        params = inspect.signature(mapping_function).parameters
+        for name, param in params.items():
+            if param.kind in [
+                inspect.Parameter.POSITIONAL_ONLY,
+                inspect.Parameter.POSITIONAL_OR_KEYWORD,
+            ]:
+                args.append(tvm.tir.Var(name, default_index_dtype))
+
+            elif param.kind == inspect.Parameter.VAR_POSITIONAL:
+                var_arg_name = name
+
+            elif param.kind == inspect.Parameter.KEYWORD_ONLY:
+                kwargs[name] = tvm.tir.Var(name, default_index_dtype)
+
+            elif param.kind in [inspect.Parameter.VAR_KEYWORD]:
+                raise ValueError("transform_layout mapping may not have **kwargs")
+
+        ndim = len(self.op.output(0).shape)
+
+        # Now that all the named arguments have been collected,
+        # everything that remains should go to the *args, if
+        # specified.
+        if var_arg_name is not None:
+            num_var_args = ndim - len(args) - len(kwargs)
+            for i in range(num_var_args):
+                args.append(tvm.tir.Var(f"{var_arg_name}[{i}]", default_index_dtype))
+
+        initial_indices = args + list(kwargs.values())
+        if len(initial_indices) != ndim:
+            raise ValueError(
+                f"transform_layout mapping accepts {len(params)} initial indices, "
+                f"but {self.op.name} is {len(self.op.shape)}-dimensional"
+            )
+
+        mapping = mapping_function(*args, **kwargs)
+
+        final_indices = []
+        axis_separators = []
+        for val in mapping:
+            if isinstance(val, tvm.ir.PrimExpr):
+                final_indices.append(val)
+            elif val is AXIS_SEPARATOR:
+                axis_separators.append(len(final_indices))
+            else:
+                raise TypeError(
+                    "Expected mapping function to return list of "
+                    "either tvm.ir.PrimExpr or tvm.te.AXIS_SEPARATOR.  "
+                    "Instead received {val} of type {type(val)}."
+                )
+
+        new_iter_vars = _ffi_api.StageTransformLayout(self, initial_indices, final_indices)
+        _ffi_api.StageSetAxisSeparators(self, axis_separators)
```

Comment:
Since axis separator is only an attribute to the buffer that affect flattening, would it be better to separate this two API calls to two schedule primitives, even if on the user-facing level? (setting axis separator doesn't have to be involved with layout transform)

Reply:
I think I'd prefer it as a single user-facing call, primarily for readability.  If the two are set in separate API calls, then there would be ambiguity in where the axis separators are placed, which would require the user to read the documentation to determine which convention is used.

```python
# Without looking at the documentation, ambiguous whether [1] sets an
# axis separator before axis 1 or after axis 1.
s[A].transform_layout(lambda i,j: [i//4, j, i%4])
s[A].set_axis_separators([1])

# Without looking at the documentation, ambiguous whether [1] sets an
# axis separator in the pre-transformation axes or the
# post-transformation axes.
s[A].set_axis_separators([1])
s[A].transform_layout(lambda i,j: [j, i//4, i%4])

# Setting both in the same user-facing API call implicitly tells the
# user that the axis separator occurs between two specific axes, and
# applies to the transformed axes.
s[A].transform_layout(lambda i,j: [i//4, j, AXIS_SEPARATOR, i%4])
```